### DISCUSSION TITLE
how to hide softkeyboard in maui

### DISCUSSION THREAD

**Discussion by persiabudi28** [dotnet/maui#4907] [Q&A] [answered]
like this:

https://theconfuzedsourcecode.wordpress.com/2017/05/19/a-keyboard-disabled-entry-control-in-xamarin-forms/

**Comment by Vpatel541** [comment]
Sorry that this isn't a good answer, but I am pretty sure its a bug

**Comment by persiabudi28** [comment]
ok, clear to the bugs for this problem.

**Reply by wsoftone** [reply]
did you clear the bug?

**Comment by matteo-convert** [comment]
After many research : 

for android
Platform from namespace Microsoft.Maui.ApplicationModel
Example in code behind event
![Capture d’écran 2022-10-19 120112](https://user-images.githubusercontent.com/93593338/196661192-d434f92b-7de4-4642-b349-29dee73edb3d.png)


     #if ANDROID

		Platform.CurrentActivity.HideKeyboard(Platform.CurrentActivity.CurrentFocus);
     #endif

Update 20 / 10 / 2022
need to check if CurrentFocus is null otherwise application crashed
![Capture d’écran 2022-10-20 144643](https://user-images.githubusercontent.com/93593338/196952658-dbf73f92-6594-439a-a39f-f8d877220de3.png)

     #if ANDROID
          if(Platform.CurrentActivity.CurrentFocus != null)
		Platform.CurrentActivity.HideKeyboard(Platform.CurrentActivity.CurrentFocus);
     #endif

**Reply by wsoftone** [reply]
the keyboard does not hide after clicking outside an Entry. why?

**Reply by matteo-convert** [reply]
didn't know i did an update today for checking if View have a focus

**Reply by sikandaramla** [reply]
Does it work with .Net 7?  I'm getting following error.
`'Activity' does not contain a definition for 'HideKeyboard' and no accessible extension method 'HideKeyboard' accepting a first argument of type 'Activity' could be found (are you missing a using directive or an assembly reference?)`

**Reply by Th3Scorpio** [reply]
Hi. Im using as follow:
Created Public Class:
 public static class KeyboardUtility
    {
        public static void HideKeyboard()
        {
            var imm = (InputMethodManager)MauiApplication.Current.GetSystemService(Android.Content.Context.InputMethodService);
            if (imm != null)
            {
                var activity = Platform.CurrentActivity;
                IBinder wToken = activity.CurrentFocus?.WindowToken;
                imm.HideSoftInputFromWindow(wToken, 0);
            }
        }

    }

then invoke from anywhere:  KeyboardUtility.HideKeyboard();

**Reply by Th3Scorpio** [reply]
For Android: 
using Microsoft.Maui.Controls;

using Android.Views.InputMethods;
using Android.OS;

**Comment by ozanyasindogan** [comment]
I was looking to be able to close the keyboard after OK click on an Entry. 
I found myself here and used the code suggested above.
It worked for me. I'm sure there should be a better way to do that but unless
someone else is also looking for it, I wanted to share my solution:

![1](https://user-images.githubusercontent.com/13081346/213801309-e314602e-6cde-4ae4-bee3-9107d982e3d8.png)
![2](https://user-images.githubusercontent.com/13081346/213801326-365f274b-17df-4f3b-9122-020c9efe1557.png)

**Comment by Rizkyuk** [comment]
Is there a to show the keyboard? Entry.Focus(); doesnt seem to work.

**Reply by ozanyasindogan** [reply]
Why don't you try Platform.CurrentActivity.ShowKeyboard, I believe it should work if this is what you need.

**Reply by Rizkyuk** [reply]
I've tried the following with no Joy
  #if ANDROID
                    Platform.CurrentActivity.ShowKeyboard(Platform.CurrentActivity.CurrentFocus);
            #endif

**Reply by melixor** [reply]
In my case Platform.CurrentActivity.CurrentFocus is always null.

**Comment by freezersharp** [comment]
What about in iOS?

**Comment by FM1973** [comment, marked as answer]
The latest Maui Community toolkit offers this, as far as I can remember.
[https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/extensions/keyboard-extensions](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/extensions/keyboard-extensions)

**Reply by Ghostbird** [reply]
The problem with that implementation is that it requires you to call it on the current active entry. In my case there's no entry on the screen at all. It's already been removed, yet the keyboard remains in both iOS and Android.

**Reply by Xyncgas** [reply]
try calling it inside removed entry

**Reply by Ghostbird** [reply]
How to call something inside something that does not exist? 

A proper separation of concerns means that I don't have any reference to anything that may, or may not, have at some point existed and caused the keyboard to be shown here.

The way we ultimately solved it, was by tacking on this functionality to the code that cleans up the entries. However from a technical perspective, the application code became worse for it. There seems no good reason that this API relies on an active entry, as is obvious from the fact that you can show the keyboard, remove the entry, and the keyboard still shows. Obviously it doesn't _need_ the entry, and the parameter ought to be optional.

**Reply by Xyncgas** [reply]
You just have to sweep the ugliness under the carpet, sometimes it's not you it is the library author's doing that makes things this way so as a consumer of the framework just say for example ```call something inside something that does not exist```, now from a higher point of view from your point of view things are abstracted away and sprinkling some procedural programming and things are pretty now

If you wanted something happen inside MAUI technically there is nothing blocking because you can always define platform specific solutions, and you are a programmer it's all 0 and 1 you can access memory from another process if you desired by exploiting Spectre/Meltdown you can do anything

**Reply by delwarthink** [reply]
does not work! I've this code but it does not hide the soft keyboard.
 txtContNum.Focus();
 txtContNum.HideKeyboardAsync(CancellationToken.None);

**Comment by eli191** [comment]
There is `HideSoftInputOnTapped="true"` option to add to the `ContentPage` to dismiss the keyboard when the user taps outside of it.

**Reply by Ghostbird** [reply]
Yes, that is a new thing in MAUI 8. Caused us a bit of trouble, because it meant that the default behaviour on iOS changed. You must now set it to get the behaviour that used to be the default. The good thing is, that when set, it now also works on Android, which is nice.

However, it is unrelated to this question, as that configures how the _user_ can hide the keyboard. This question was about how _we developers_ can hide the keyboard.

**Reply by davidbritch** [reply]
https://learn.microsoft.com/dotnet/maui/user-interface/controls/entry?view=net-maui-8.0#hide-and-show-the-soft-input-keyboard

**Reply by Ghostbird** [reply]
This is the `Entry` API. See the accepted answer thread for the problem with that.

**Reply by CodingOctocat** [reply]
HideSoftInputOnTapped is bugged, for Editor, when clicking inside Editor (e.g. to change cursor position), the keyboard will be hidden, when clicking again, the keyboard will be shown.

**Comment by tjiakwokyung28** [comment]
how to run on ios

**Comment by mohmmadkhalil** [comment]
Simply add this line to the ContentPage definition, and it will solve the problem:

HideSoftInputOnTapped="True"

Like this

<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="Alundi.App.Views.OTPPage"
             xmlns:mi="http://www.aathifmahir.com/dotnet/2022/maui/icons"
             xmlns:resx="clr-namespace:Alundi.App.Resources.Languages"
             **HideSoftInputOnTapped="True"**
